src/Payments.sol

```diff
@@ -0,0 +1,793 @@
+                );
+
+            // If arbiter returned no progress, exit early without updating state
+            if (rail.settledUpTo <= processedEpoch) {
+                return (totalSettled, rail.settledUpTo, arbitrationNote);
+            }
+
+            // Add the settled amount to our running total
+            totalSettled += segmentAmount;
+
+            // If arbiter partially settled the segment, exit early
+            if (rail.settledUpTo < segmentEndBoundary) {
+                return (totalSettled, rail.settledUpTo, arbitrationNote);
+            }
+
+            // Successfully settled full segment, update tracking values
+            processedEpoch = rail.settledUpTo;
+            note = arbitrationNote;
+
+            // Remove the processed rate change from the queue
+            if (!rateQueue.isEmpty()) {
+                rateQueue.dequeue();
+            }
+        }
+
+        // We've successfully settled up to the target epoch
+        return (totalSettled, rail.settledUpTo, note);
+    }
+
+    function _settleSegment(
+        uint256 railId,
+        uint256 epochStart,
+        uint256 epochEnd,
+        uint256 rate
+    ) internal returns (uint256 settledAmount, string memory note) {
+        Rail storage rail = rails[railId];
+        Account storage payer = accounts[rail.token][rail.from];
+        Account storage payee = accounts[rail.token][rail.to];
+
+        // Calculate the default settlement values (without arbitration)
+        uint256 duration = epochEnd - epochStart;
+        settledAmount = rate * duration;
+        uint256 settledUntilEpoch = epochEnd;
+        note = "";
+
+        // If this rail has an arbiter, let it decide on the final settlement amount
+        if (rail.arbiter != address(0)) {
+            IArbiter arbiter = IArbiter(rail.arbiter);
+            IArbiter.ArbitrationResult memory result = arbiter.arbitratePayment(
+                railId,
+                settledAmount,
+                epochStart,
+                epochEnd
+            );
+
+            // Ensure arbiter doesn't settle beyond our segment's end boundary
+            require(
+                result.settleUpto <= epochEnd,
+                "arbiter settled beyond segment end"
+            );
+
+            settledUntilEpoch = result.settleUpto;
+            settledAmount = result.modifiedAmount;
+            note = result.note;
+
+            // If arbiter made no progress, return early without updating any state
+            if (settledUntilEpoch <= epochStart) {
+                return (0, note);
+            }
+
+            // Ensure arbiter doesn't allow more payment than the maximum possible
+            // for the epochs they're confirming
+            uint256 maxAllowedAmount = rate * (settledUntilEpoch - epochStart);
+            require(
+                result.modifiedAmount <= maxAllowedAmount,
+                "arbiter modified amount exceeds maximum for settled duration"
+            );
```

Comment:
I'd:

1. Require that `settledUntilEpoch >= epochStart`.
2. Require that the amount settled is within expectations given the `settledUntilEpoch`.
3. Finally, short-circuit if no funds were actually settled.  Or, TBH, I'd just go to the end of the function anyways and not short-circuit.

Specifically, I don't want to return `(0, note)` if the arbiter returns nonsensical results (e.g., settle 1M FIL up till some epoch in the past).

Reply:
This is done.

Reply:
TODO for @aarshkshah1992 : Audit the code for off by ones to enforce that the `untilEpoch` in the rate-change queue is not inclusive of the end epoch and that `rail.settledUpto` is inclusive of the settledEpoch.

Reply:
Will do this in the follow-up PR.